OAK-9880 : simplify rgc query

stefan-egli opened the pull request.

stefan-egli committed:
OAK-9880 : simplify rgc query

stefan-egli committed:
OAK-9880 : method renamed to reflect new return type

stefan-egli commented:
one concern that just came to mind with this: obviously this would increase the number of queries executed - and that scales with the number of clusterNodeIds in use...

stefan-egli commented:
Maybe to compensate for that, do less frequent RGC, currently it runs every 5sec by default - changing that to every 1min might not have too much of an impact

mreutegg commented:
> one concern that just came to mind with this: obviously this would increase the number of queries executed - and that scales with the number of clusterNodeIds in use...

We could address this to some degree by remembering successful operations on minMaxRevTimeInSecs per clusterId. An inactive cluster node will not have its sweep revision updated and minMaxRevTimeInSecs will not change. This would limit operations to the number of active clusterIds.

> do less frequent RGC, currently it runs every 5sec by default - changing that to every 1min might not have too much of an impact

I would prefer to keep the interval small. This ensures RGC activity is stretched out more evenly and avoids spikes that impact an application. Every 5 seconds what an initial default set for this kind of continuous RGC. We can certainly discuss a change. Maybe in a separate issue?

mreutegg reviewed (changes requested):
Changes look good to me. This change does not come with a test. Could you at least add a test that checks an expected number of deleteMany operations are issued for a given set of clusterIds?

stefan-egli committed:
OAK-9880 : test case added

stefan-egli committed:
Merge branch 'apache:trunk' into OAK-9880

stefan-egli committed:
OAK-9880 : added lastDefaultNoBranchDeletionRevs to MongoVersionGCSupport - plus more tests. This will now avoid doing queries for idle clusterNodeIds , something which is more relevant now that we split one of the queries into several, one per clusterNodeId

stefan-egli committed:
OAK-9880 : more asserts to count number of deleteMany calls

stefan-egli commented:
* added logic to avoid calling `deleteMany` for idle clusterNodeIds - plus tests for this
* added a test for number of `deleteMany` calls for the above and some of the existing tests

stefan-egli commented:
@mreutegg, as I have modified this PR from your earlier review would appreciate a second (final?) review, thx!

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoVersionGCSupport.java:
```diff
@@ -100,6 +105,11 @@ public MongoVersionGCSupport(MongoDocumentStore store) {
     }
 
+    /** accessor for testing only */
+    protected RevisionVector getLastDefaultNoBranchDeletionRevs() {
```
I cannot find any usages of this method. Remove?

stefan-egli replied:
agree, a left over from a previous test version. removed now

mreutegg commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MongoVersionGCSupportDefaultNoBranchTest.java:
```diff
@@ -0,0 +1,431 @@
+                        connections.add(connection);
+                        MongoDatabase db = connection.getDatabase();
+                        final AtomicReference<String> noEx = new AtomicReference<>(null);
+                        MongoTestDatabase testDb = new MongoTestDatabase(db, noEx, noEx, noEx) {
```
Could this also go to `CountingMongoDatabase` and `CountingMongoCollection`?

stefan-egli replied:
yes, can (and without interfering with existing tests). done now in https://github.com/apache/jackrabbit-oak/pull/651/commits/d82ba32226e09ec545e348bc57367167052bfdbb

mreutegg reviewed (approved):
Other than above, looks good to me.

stefan-egli committed:
OAK-9880 : remove unused/orphaned method

stefan-egli committed:
Merge branch 'apache:trunk' into OAK-9880

stefan-egli committed:
OAK-9880 : moved deleteMany counters to CountingMongoCollection
